{
  "cve_id": "CVE-2024-26130",
  "github_data": {
    "repository": "pyca/cryptography",
    "fix_commit": "97d231672763cdb5959a3b191e692a362f1b9e55",
    "related_commits": [
      "97d231672763cdb5959a3b191e692a362f1b9e55",
      "97d231672763cdb5959a3b191e692a362f1b9e55"
    ],
    "patch_url": "https://github.com/pyca/cryptography/commit/97d231672763cdb5959a3b191e692a362f1b9e55.patch",
    "fix_commit_details": {
      "sha": "97d231672763cdb5959a3b191e692a362f1b9e55",
      "commit_date": "2024-02-19T16:50:28Z",
      "author": {
        "login": "alex",
        "type": "User",
        "stats": {
          "total_commits": 4529,
          "average_weekly_commits": 7.573578595317726,
          "total_additions": 0,
          "total_deletions": 0,
          "weeks_active": 507
        }
      },
      "commit_message": {
        "title": "Fixes #10422 -- don't crash when a PKCS#12 key and cert don't match (#10423)",
        "length": 76,
        "has_description": false,
        "references_issue": true
      },
      "stats": {
        "total": 27,
        "additions": 27,
        "deletions": 0
      },
      "files": [
        {
          "filename": "src/cryptography/hazmat/backends/openssl/backend.py",
          "status": "modified",
          "additions": 9,
          "deletions": 0,
          "patch": "@@ -623,6 +623,15 @@ def serialize_key_and_certificates_to_pkcs12(\n                     mac_iter,\n                     0,\n                 )\n+                if p12 == self._ffi.NULL:\n+                    errors = self._consume_errors()\n+                    raise ValueError(\n+                        (\n+                            \"Failed to create PKCS12 (does the key match the \"\n+                            \"certificate?)\"\n+                        ),\n+                        errors,\n+                    )\n \n             if (\n                 self._lib.Cryptography_HAS_PKCS12_SET_MAC"
        },
        {
          "filename": "tests/hazmat/primitives/test_pkcs12.py",
          "status": "modified",
          "additions": 18,
          "deletions": 0,
          "patch": "@@ -660,6 +660,24 @@ def test_key_serialization_encryption_set_mac_unsupported(\n                 b\"name\", cakey, cacert, [], algorithm\n             )\n \n+    @pytest.mark.supported(\n+        only_if=lambda backend: backend._lib.Cryptography_HAS_PKCS12_SET_MAC,\n+        skip_message=\"Requires OpenSSL with PKCS12_set_mac\",\n+    )\n+    def test_set_mac_key_certificate_mismatch(self, backend):\n+        cacert, _ = _load_ca(backend)\n+        key = ec.generate_private_key(ec.SECP256R1())\n+        encryption = (\n+            serialization.PrivateFormat.PKCS12.encryption_builder()\n+            .hmac_hash(hashes.SHA256())\n+            .build(b\"password\")\n+        )\n+\n+        with pytest.raises(ValueError):\n+            serialize_key_and_certificates(\n+                b\"name\", key, cacert, [], encryption\n+            )\n+\n \n @pytest.mark.skip_fips(\n     reason=\"PKCS12 unsupported in FIPS mode. So much bad crypto in it.\""
        }
      ],
      "file_patterns": {
        "security_files": 1,
        "config_files": 0,
        "dependency_files": 0,
        "test_files": 1,
        "unique_directories": 2,
        "max_directory_depth": 5
      },
      "context": {
        "surrounding_commits": [
          {
            "sha": "b74d321d16c0927d5c705104644ad0addd762e5a",
            "date": "2025-01-14T00:39:01Z",
            "author_login": "pyca-boringbot[bot]"
          },
          {
            "sha": "b15591e93af508bf3fa9af51e94d65a46fe72d82",
            "date": "2025-01-14T00:36:50Z",
            "author_login": "pyca-boringbot[bot]"
          },
          {
            "sha": "78d354711b20e6e6cb23ca5879cae7030d46673a",
            "date": "2025-01-13T14:44:59Z",
            "author_login": "alex"
          },
          {
            "sha": "b17c77b5d730df63238a30a456a7d18cb50cc1af",
            "date": "2025-01-13T12:49:21Z",
            "author_login": "dependabot[bot]"
          },
          {
            "sha": "cf8ec124724c557bd55735fc72035c53030b7481",
            "date": "2025-01-13T12:48:54Z",
            "author_login": "dependabot[bot]"
          }
        ]
      }
    }
  },
  "vulnerability_details": {
    "cvss_score": 7.5,
    "cvss_vector": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H",
    "cwe_id": "CWE-476",
    "description": "cryptography is a package designed to expose cryptographic primitives and recipes to Python developers. Starting in version 38.0.0 and prior to version 42.0.4, if `pkcs12.serialize_key_and_certificates` is called with both a certificate whose public key did not match the provided private key and an `encryption_algorithm` with `hmac_hash` set (via `PrivateFormat.PKCS12.encryption_builder().hmac_hash(...)`, then a NULL pointer dereference would occur, crashing the Python process. This has been resolved in version 42.0.4, the first version in which a `ValueError` is properly raised.",
    "attack_vector": "NETWORK",
    "attack_complexity": "LOW"
  },
  "temporal_data": {
    "published_date": "2024-02-21T17:15:09.863",
    "last_modified": "2024-11-21T09:01:59.757",
    "fix_date": "2024-02-19T16:50:28Z"
  },
  "references": [
    {
      "url": "https://github.com/pyca/cryptography/commit/97d231672763cdb5959a3b191e692a362f1b9e55",
      "source": "security-advisories@github.com",
      "tags": []
    },
    {
      "url": "https://github.com/pyca/cryptography/pull/10423",
      "source": "security-advisories@github.com",
      "tags": []
    },
    {
      "url": "https://github.com/pyca/cryptography/security/advisories/GHSA-6vqw-3v5j-54x4",
      "source": "security-advisories@github.com",
      "tags": []
    },
    {
      "url": "https://github.com/pyca/cryptography/commit/97d231672763cdb5959a3b191e692a362f1b9e55",
      "source": "af854a3a-2127-422b-91ae-364da2661108",
      "tags": []
    },
    {
      "url": "https://github.com/pyca/cryptography/pull/10423",
      "source": "af854a3a-2127-422b-91ae-364da2661108",
      "tags": []
    },
    {
      "url": "https://github.com/pyca/cryptography/security/advisories/GHSA-6vqw-3v5j-54x4",
      "source": "af854a3a-2127-422b-91ae-364da2661108",
      "tags": []
    }
  ],
  "collection_metadata": {
    "collected_at": "2025-01-11T23:07:28.142501",
    "processing_status": "raw"
  },
  "repository_context": {
    "name": "cryptography",
    "owner": "pyca",
    "created_at": "2013-08-07T02:23:38Z",
    "updated_at": "2025-01-14T00:39:06Z",
    "pushed_at": "2025-01-14T00:39:02Z",
    "size": 57765,
    "stars": 6794,
    "forks": 1553,
    "open_issues": 51,
    "watchers": 6794,
    "has_security_policy": false,
    "default_branch": "main",
    "protected_branches": [
      "0.2.x",
      "0.5.x",
      "0.6.x",
      "0.7.x",
      "0.8.x",
      "0.9.x",
      "1.0.x",
      "1.1.x",
      "1.2.x",
      "1.3.x",
      "1.5.x",
      "1.7.x",
      "1.8.x",
      "2.0.x",
      "2.1.x",
      "2.2.x",
      "2.3.x",
      "2.4.x",
      "2.9.x",
      "3.1.x",
      "3.2.x",
      "3.3.x",
      "3.4.x",
      "36.0.x",
      "37.0.x",
      "38.0.x",
      "39.0.x",
      "40.0.x",
      "41.0.x",
      "42.0.x"
    ],
    "languages": {
      "Python": 2248082,
      "Rust": 816136,
      "Shell": 1558
    },
    "commit_activity": {
      "total_commits_last_year": 1726,
      "avg_commits_per_week": 33.19230769230769,
      "days_active_last_year": 341
    },
    "security_features": {
      "has_security_policy": false,
      "has_protected_branches": true,
      "has_wiki": false,
      "has_issues": true,
      "allow_forking": true,
      "is_template": false,
      "license": "other"
    },
    "collected_at": "2025-01-14T12:53:50.605682"
  }
}